### ISSUE TITLE
BG/NBD Model in PyMC

### ISSUE THREAD

**Issue by juanitorduz** [ColtAllen/lifetimes#6]
Hey! I have recently drafted a BG/NBD model in PyMC, see https://juanitorduz.github.io/bg_nbd_pymc/
In particular I have also added the possibility to add time-invariant regressors (even though the sampler and/or model parametrisation need some improvements). Other models from the `lifetimes` package can be ported to PyMC in a similar way (because the `lifetimes` package has all the log-likelihood functions written in `numpy` already).

**Comment by ColtAllen** [comment]
Hey @juanitorduz ,

This is great! I really hope you can join the Zoom kickoff call for this project next Sunday. Information to join is in a Discussion post in this repo as well as here:

https://github.com/CamDavidsonPilon/lifetimes/issues/414#issuecomment-1073247582

I've been working on a BG/NBD model as well, but will be experimenting with ADVI or SVGD for inference because MCMC training can go on for days, compared to minutes with ADVI. However, MCMC is also the only way to derive the true posterior distribution (rather than an approximation as is the case with ADVI or SVGD). A more performant option for MCMC may be Stan, which `prophet` uses for inference of its time series model parameters. @alexpavlakis wrote a script in Stan for the Gamma-Gamma model, but I haven't tested it yet in `CmndStanPy`.

As you mentioned, the tricky thing about these algorithms is choosing an appropriate prior, and my goal is to make this library as user-friendly as possible. Fortunately in the case of the Pareto/NBD model, there is a research paper on this very topic:

[Worth the effort? Comparison of different MCMC algorithms for estimating the Pareto/NBD model](https://link.springer.com/article/10.1007/s11573-021-01057-6)

I've saved several research papers on covariates and other modeling alternatives and will be posting links to the README soon. None of them are pay-walled, but to be safe I can also look into uploading my saved copies to the repo. One of these papers even covers a Pareto/NBD model with time-varying covariates, but it is immensely more compute intensive than the standard Pareto/NBD and only becomes more performant for customers with lifetimes of 3 years or more.

**Comment by juanitorduz** [comment]
Thanks for the input! I will try to attend the zoom call :)

**Comment by ColtAllen** [comment]
@juanitorduz I just wanted to let you know I've been refactoring the code in your notebook to run in `pymc4` and `aesara`. Transitioning from the existing MLE backend estimations to `pymc` is going to require such a significant foundational restructuring that I intend to rebrand it once it is ready to be published to PyPI. For now all pertinent future work branches shall be created under the  `pymc` sub-branch.

**Comment by juanitorduz** [comment]
That sounds very exiting! Let me know if you need som input or want me to test some of the ongoing features. As a side remark, I also did a small re-write in `pymc` for the gamma-gamma model, see https://juanitorduz.github.io/gamma_gamma_pymc/

**Comment by ColtAllen** [comment]
> That sounds very exiting! Let me know if you need some input or want me to test some of the ongoing features.

Thanks! The likelihood function for the hyperpriors is raising an exception in v4. I rewrote it per [this paper](http://brucehardie.com/notes/027/bgnbd_num_error.pdf) but it still persists. I just created an issue for it:

https://github.com/pymc-devs/pymc/issues/5730

**Comment by ColtAllen** [comment]
@juanitorduz the PR for the adaptation of your notebook code into `lifetimes` is now up!

https://github.com/ColtAllen/lifetimes/pull/24
